primitives/consensus/beefy/src/lib.rs

```diff
@@ -104,6 +98,28 @@ pub mod crypto {
 	}
 }
 
+/// BEEFY cryptographic types for BLS crypto
+///
+/// This module basically introduces three crypto types:
+/// - `bls_crypto::Pair`
+/// - `bls_crypto::Public`
+/// - `bls_crypto::Signature`
+///
+/// Your code should use the above types as concrete types for all crypto related
+/// functionality.
+
+#[cfg(feature = "bls-experimental")]
+pub mod bls_crypto {
+	use sp_application_crypto::{app_crypto, bls377, key_types::BEEFY as BEEFY_KEY_TYPE};
+	app_crypto!(bls377, BEEFY_KEY_TYPE);
+
+	/// Identity of a BEEFY authority using BLS as its crypto.
+	pub type AuthorityId = Public;
+
+	/// Signature for a BEEFY authority using BLS as its crypto.
+	pub type AuthoritySignature = Signature;
+}
```

Comment:
Ditto. And verifying (ECDSA, BLS) signature is different as we only verifying ECDSA in gossip.

But I wonder if it is better the way it is done now in #13311 and make the BEEFY keystore generic over AuthId or we should actually define `pub mod ecdsa_n_bls_crypto` in `primitives/consensu/beefy` instead.

What do you think?

Reply:
My suggestion does not work out of the box because `type AuthorityId = (ecdsa_crypto::Public, bls_crypto::Public)` does not implement `RuntimeAppPublic` But we still can have mod for this type of `AuthorityId` and use it in client. We'll resolve this restricton in future improvement so we can implement BeefyAuthorityId for `(ecdsa_crypto::Public, bls_crypto::Public)`

Reply:
> But can't we just call the Pair::verify so that we implement the BeefyAuthorityId over a generic AuthorityId<P> where P: Pair?

The reason they do not call `Pair::verify` beause they want to have control over the hasher which I believe Pair::verify doesn't give.

Reply:
I also have to re-introduce `w3f-bls` as optional dependancy to implement `BeefyAuthorityId` for `bls_crypto::AuthorityId` because we do not have a host call for verifying bls signatures.